### PR TITLE
HW4: use PostgreSQL in AWS

### PR DESCRIPTION
**Pull request by candentira**
1. Task: https://github.com/rolling-scopes-school/nodejs-aws-tasks/blob/main/task4-rds/task.md
2. Screenshot:
image
image
3. Links: 
getProductsList: GET https://dnfpunajrc.execute-api.eu-west-1.amazonaws.com/dev/products
createProducts: POST https://dnfpunajrc.execute-api.eu-west-1.amazonaws.com/dev/products
getProductsById: GET https://dnfpunajrc.execute-api.eu-west-1.amazonaws.com/dev/products/{id}
FE: https://d5vx1ujajushw.cloudfront.net/
FE PR: https://github.com/candentira/nodejs-aws-fe/pull/3
4. Done 08.11.2020 / deadline 08.11.2020
5. Score: 9 / 9

EVALUATION CRITERIA:

Reviewers should verify the lambda functions by invoking them through provided URLs.
- [x] 1 - Task 4.1 is implemented
- [x] 3 - TASK 4.2 is implemented lambda links are provided and returns data
- [x] 4 - TASK 4.3 is implemented lambda links are provided and products is stored in DB (call TASK 4.2 to see the product)
- [x] 5 - Your own Frontend application is integrated with product service (/products API) and products from product-service are represented on Frontend. Link to a working Front-End application is provided for cross-check reviewer.
Additional (optional) tasks (but nice to have):

- [x] +1 - POST/products lambda functions returns error 400 status code if product data is invalid
- [x] +1 - All lambdas return error 500 status code on any error (DB connection, any unhandled error in code)
- [x] +1 - All lambdas do console.log for each incoming requests and their arguments
- [x] +1 - Transaction based creation of product (in case stock creation is failed then related to this stock product is not created and not ready to be used by the end user and vice versa) (https://devcenter.kinvey.com/nodejs/tutorials/bl-transactional-support)

### CODE REVIEW

File: product-service/handlers/createProducts.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        }
+    }
+
+    await createProducts({ title, description, price, count });
+    
+    console.info("createProducts Lambda finished execution successfully");
+
+    return {
+      ...responseHeaders,
+      statusCode: 200,
```
Comment (ahordzin) [inline product-service/handlers/createProducts.ts]: Please change statusCode to 201.

File: product-service/handlers/services/productsService.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+export const createProducts = async ({ title, description, price, count }) => {
+  const client = new Client(dbOptions);
+  await client.connect();
+  try {
+    await client.query('BEGIN');
+    const createProductQuery = `INSERT INTO products (title, description, price) VALUES ($1, $2, $3) RETURNING id`
+    const result = await client.query(createProductQuery, [title, description, price]) ;
+    await client.query(`INSERT INTO stocks (product_id, count) VALUES ($1, $2)`, [result.rows[0].id, count]);
+    await client.query('COMMIT');
```
Comment (ahordzin) [inline product-service/handlers/services/productsService.ts:55]: The create response shouldn't be empty. Please add at least product id to the response

File: product-service/handlers/createProducts.ts (hunk 1 above)
Comment (candentira) [reply product-service/handlers/createProducts.ts]: Done

File: product-service/handlers/services/productsService.ts (hunk 2 above)
Comment (candentira) [reply product-service/handlers/services/productsService.ts:55]: Now am returning created resource.

**Pull request merged**